support/BRFileService.c

```diff
@@ -127,34 +228,158 @@ fileServiceCreate (const char *basePath,
     if (NULL == dir) return NULL;
     closedir(dir);
 
-    // Create the directory hierarchy
-    size_t pathToTypeSize = strlen(basePath) + 1 + strlen(currency) + 1 + strlen(network) + 1;
-    char dirPath[pathToTypeSize];
-
-    sprintf (dirPath, "%s/%s", basePath, currency);
-    if (-1 == directoryMake(dirPath)) return NULL;
-
-    sprintf(dirPath, "%s/%s/%s", basePath, currency, network);
-    if (-1 == directoryMake(dirPath)) return NULL;
+    // Require SQLite to support 'MULTI_THREADED' or 'SERIALIZED'.  We'll lock our connection.
+    // and thus 'MULTI_THREADED' is appropriate.
+    if (0 == sqlite3_threadsafe()) return NULL;
 
+    // Create the file service itself
     BRFileService fs = calloc (1, sizeof (struct BRFileServiceRecord));
 
-    fs->pathToType = strdup(dirPath);
-    array_new (fs->entityTypes, FILE_SERVICE_INITIAL_TYPE_COUNT);
+    // A recursive lock - fileServiceLoad() will recursively call fileServiceSave()
+    {
+        pthread_mutexattr_t attr;
+        pthread_mutexattr_init(&attr);
+        pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_RECURSIVE);
```

Comment:
Recursive locks are a bit controversial. Would it be hard to remove its usage here? What is your opinion of them?

Reply:
There is a recursive call to `fileServiceSave()`; you'd say avoid the RECURSIVE lock by splitting `fileServiceSave()` into a lock-free 'static' internal function?

Reply:
Comes down to if we want to continue using recursive locks; if yes, no issue, if no, splitting would work.

Reply:
Added `int needLock` parameter to a static version as `_fileServiceSave()`